state/protocol/protocol_state/kvstore.go

```diff
@@ -0,0 +1,108 @@
+// for this software version.
+type KVStoreReader interface {
+	ID() flow.Identifier
+
+	// v0
+
+	VersionedEncodable
+
+	// GetProtocolStateVersion returns the current Protocol State version.
+	// The Protocol State version specifies the current version of the Protocol State,
+	// including the key-value store. Changes in the protocol state version
+	// correspond to changes in the set of key-value pairs which are supported,
+	// and which model is used for serialization.
+	// It can be updated by an UpdateKVStoreVersion service event.
+	GetProtocolStateVersion() uint64
+
+	// GetVersionUpgrade returns the upgrade version of protocol.
+	// VersionUpgrade is a view-based activator that specifies the version which has to be applied
+	// and the view at which it has to be applied. It may contain an old value if the upgrade has already applied.
+	// It can be updated by an flow.ProtocolStateVersionUpgrade service event.
+	GetVersionUpgrade() *ViewBasedActivator[uint64]
+
+	// v1
+
+	// GetInvalidEpochTransitionAttempted returns a flag indicating whether epoch
+	// fallback mode has been tentatively triggered on this fork.
+	// Errors:
+	//  - ErrKeyNotSupported if the key does not exist in the current model version.
+	GetInvalidEpochTransitionAttempted() (bool, error)
+}
+
+// KVStoreAPI is the latest read-writer interface to the Protocol State key-value store.
+// It is backward-compatible with all versioned model types defined in the models.go
+// for this software version.
+type KVStoreAPI interface {
+	KVStoreReader
+
+	// Clone returns a deep copy of the KVStoreAPI.
+	// This is used to create a new instance of the KVStoreAPI to avoid mutating the original.
+	Clone() KVStoreAPI
+
+	// v0
+
+	// SetVersionUpgrade sets the protocol upgrade version. This method is used
+	// to update the Protocol State version when a flow.ProtocolStateVersionUpgrade is processed.
+	// It contains the new version and the view at which it has to be applied.
+	SetVersionUpgrade(version *ViewBasedActivator[uint64])
+
+	// v1
+
+	// SetInvalidEpochTransitionAttempted sets the epoch fallback mode flag.
+	// Errors:
+	//  - ErrKeyNotSupported if the key does not exist in the current model version.
+	SetInvalidEpochTransitionAttempted(attempted bool) error
+}
+
+// KeyValueStoreStateMachine implements a low-level interface for state-changing operations on the key-value store.
+// It is used by higher level logic to evolve the protocol state when certain events that are stored in blocks are observed.
+// The KeyValueStoreStateMachine is stateful and internally tracks the current state of key-value store.
+// A separate instance is created for each block that is being processed.
+type KeyValueStoreStateMachine interface {
+	// Build returns updated key-value store model, state ID and a flag indicating if there were any changes.
+	Build() (updatedState KVStoreReader, stateID flow.Identifier, hasChanges bool)
```

Comment:
In my mind, it would be great if the `Build` method returned a representation of the protocol state snapshot that permits mutations. So specifically `KVStoreAPI`. For instance, if we cache the results to apply the state-changing events from the next block (one we receive it), we would want them represented via the mutable `KVStoreAPI`  (hope that makes sense).

```suggestion
	Build() (updatedState KVStoreAPI, snapshotID flow.Identifier, hasChanges bool)
```